File: src/test/resources/sample_module/io/github/grimch/doclet/sample_module/types/basic/I1.java
/*
 * This file is part of LogiDoclet.
 *
 * Copyright (c) 2025 The LogiDoclet Authors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package io.github.grimch.doclet.sample_module.types.basic;

// *** Interface with Generics and Wildcards ***
// Note: This interface is now non-sealed to participate in the sealed hierarchy test.
/**
 * Represents a sealed interface demonstrating generics, wildcards, default, and static methods.
 * This interface is part of a sealed hierarchy, permitting {@link C1} and {@link io.github.grimch.doclet.sample_module.types.advanced.C4}
 * as direct implementations.
 *
 * @param <T> The first type parameter, which must extend {@link Number}.
 * @param <U> The second type parameter.
 * @see C1
 * @see io.github.grimch.doclet.sample_module.types.advanced.C4
 */
public sealed interface I1<T extends Number, U> permits C1, io.github.grimch.doclet.sample_module.types.advanced.C4 {
    /**
     * An abstract method that takes an argument of type {@code T}.
     * @param arg1 The argument of type {@code T}.
     */
    void m1(T arg1);
    /**
     * A default method that returns its input argument.
     * @param arg2 The argument of type {@code U}.
     * @return The input argument {@code arg2}.
     */
    default U m2(U arg2) { return arg2; }
    /**
     * A static generic method that prints its argument to the console.
     * @param <V> The type parameter for this method.
     * @param arg3 The argument of type {@code V}.
     */
    static <V> void m3(V arg3) { System.out.println(arg3); }
}

File: src/test/resources/expected_output/full/io/github/grimch/doclet/sample_module/types/basic/I1.pl
interface(
    'I1',
    'io.github.grimch.doclet.sample_module.types.basic',
    [
        modifier(public),
        modifier(abstract),
        modifier(sealed)
    ],
    [
        type_parameter(
            'T',
            [
                declared_type('java.lang.Number', [])
            ],
            []
        ),
        type_parameter(
            'U',
            [
                declared_type('java.lang.Object', [])
            ],
            []
        )
    ],
    [],
    [
        method(
            m1,
            [
                modifier(public),
                modifier(abstract)
            ],
            [],
            type(no_type, void),
            [
                parameter(
                    arg1,
                    type(type_variable, 'T'),
                    [],
                    []
                )
            ],
            [],
            [],
            ' An abstract method that takes an argument of type {@code T}.\n @param arg1 The argument of type {@code T}.\n'
        ),
        method(
            m2,
            [
                modifier(public),
                modifier(default)
            ],
            [],
            type(type_variable, 'U'),
            [
                parameter(
                    arg2,
                    type(type_variable, 'U'),
                    [],
                    []
                )
            ],
            [],
            [],
            ' A default method that returns its input argument.\n @param arg2 The argument of type {@code U}.\n @return The input argument {@code arg2}.\n'
        ),
        method(
            m3,
            [
                modifier(public),
                modifier(static)
            ],
            [
                type_parameter(
                    'V',
                    [
                        declared_type('java.lang.Object', [])
                    ],
                    []
                )
            ],
            type(no_type, void),
            [
                parameter(
                    arg3,
                    type(type_variable, 'V'),
                    [],
                    []
                )
            ],
            [],
            [],
            ' A static generic method that prints its argument to the console.\n @param <V> The type parameter for this method.\n @param arg3 The argument of type {@code V}.\n'
        )
    ],
    [],
    [
        declared_type('io.github.grimch.doclet.sample_module.types.basic.C1', []),
        declared_type('io.github.grimch.doclet.sample_module.types.advanced.C4', [])
    ],
    ' Represents a sealed interface demonstrating generics, wildcards, default, and static methods.\n This interface is part of a sealed hierarchy, permitting {@link C1} and {@link io.github.grimch.doclet.sample_module.types.advanced.C4}\n as direct implementations.\n\n @param <T> The first type parameter, which must extend {@link Number}.\n @param <U> The second type parameter.\n @see C1\n @see io.github.grimch.doclet.sample_module.types.advanced.C4\n'
).
